ext/opcache/jit/zend_jit_vm_helpers.c

```diff
@@ -424,31 +479,55 @@ static zend_always_inline ZEND_OPCODE_HANDLER_RET ZEND_FASTCALL zend_jit_trace_c
 		opline = execute_data ? EX(opline) : NULL;
 #ifdef HAVE_GCC_GLOBAL_REGS
 		return;
+#elif ZEND_VM_TAIL_CALL_DISPATCH
+		if (EXPECTED(opline)) {
+			ZEND_OPCODE_TAIL_CALL((zend_vm_opcode_handler_t)opline->handler);
+		} else {
+			return (const zend_op*)ZEND_VM_ENTER_BIT;
+		}
 #else
 		return (const zend_op*)((uintptr_t)opline | ZEND_VM_ENTER_BIT); // ZEND_VM_ENTER() / ZEND_VM_RETURN()
 #endif
 	} else {
-		zend_vm_opcode_handler_t handler = (zend_vm_opcode_handler_t)ZEND_OP_TRACE_INFO(opline, offset)->orig_handler;
+		zend_vm_opcode_handler_t handler = ZEND_OP_TRACE_INFO(opline, offset)->orig_handler;
 		ZEND_OPCODE_TAIL_CALL(handler);
 	}
 }
 
-ZEND_OPCODE_HANDLER_RET ZEND_FASTCALL zend_jit_func_trace_helper(ZEND_OPCODE_HANDLER_ARGS)
+ZEND_OPCODE_HANDLER_RET ZEND_OPCODE_HANDLER_CCONV zend_jit_func_trace_helper(ZEND_OPCODE_HANDLER_ARGS)
 {
-	ZEND_OPCODE_TAIL_CALL_EX(zend_jit_trace_counter_helper,
-		((ZEND_JIT_COUNTER_INIT + JIT_G(hot_func) - 1) / JIT_G(hot_func)));
+	zend_jit_op_array_trace_extension *jit_extension =
+		(zend_jit_op_array_trace_extension*)ZEND_FUNC_INFO(&EX(func)->op_array);
+	size_t offset = jit_extension->offset;
+	uint32_t cost = ((ZEND_JIT_COUNTER_INIT + JIT_G(hot_func) - 1) / JIT_G(hot_func));
+
+	*(ZEND_OP_TRACE_INFO(opline, offset)->counter) -= cost;
+
+	ZEND_OPCODE_TAIL_CALL(zend_jit_trace_counter_helper);
```

Comment:
Tail-calling was not possible due to the extra arg in `zend_jit_trace_counter_helper`, so I removed it.

Alternative would be to define these handlers in IR, as we do for the hybrid JIT.